infoblox/provider.go

```diff
@@ -21,36 +21,74 @@ const (
 	altIdSeparator      = "|"
 )
 
-func generateInternalId() string {
-	return uuid.NewString()
+type internalResourceId struct {
+	value uuid.UUID
 }
 
-func generateAltId(internalId string, ref string) string {
-	return fmt.Sprintf(
-		"%s%s%s",
-		internalId, altIdSeparator, ref)
+func (id *internalResourceId) Equal(id2 *internalResourceId) bool {
+	if id2 == nil {
+		panic("the argument must not be nil")
+	}
+	return id.value.String() == id2.value.String()
 }
 
-func getAltIdFields(altId string) (
-	internalId string, ref string, err error) {
+func (id *internalResourceId) String() string {
+	return id.value.String()
+}
 
-	idParts := strings.SplitN(altId, altIdSeparator, 2)
-	if len(idParts) != 2 {
-		err = fmt.Errorf("invalid internal ID for host record: '%s'", altId)
-		return
+// Returns a pointer to parsed internal resource ID, nil otherwise.
+func newInternalResourceIdFromString(id string) *internalResourceId {
+	newUUID, err := uuid.Parse(id)
+	if err != nil {
+		return nil
 	}
 
-	internalId = idParts[0]
-	ref = idParts[1]
-	return
+	return &internalResourceId{value: newUUID}
 }
 
-func renewAltId(oldAltId string, newRef string) (string, error) {
-	internalId, _, err := getAltIdFields(oldAltId)
+func generateInternalId() *internalResourceId {
+	uuid, err := uuid.NewRandom()
 	if err != nil {
-		return "", err
+		panic(err)
 	}
-	return generateAltId(internalId, newRef), nil
+	return &internalResourceId{value: uuid}
+}
+
+func isValidInternalId(internalId string) bool {
+	_, err := uuid.Parse(internalId)
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+func generateAltId(internalId *internalResourceId, ref string) string {
+	if internalId == nil {
+		panic("the argument must not be nil")
+	}
+	return fmt.Sprintf(
+		"%s%s%s",
+		internalId.String(), altIdSeparator, ref)
+}
+
+// valid = true:
+//   - exactly 2 parts found
+//   - ... and separated by the delimiter
+//   - ... and the 1st one is a valid internal ID
+//   - ... and the 2nd one is not empty
+func getAltIdFields(altId string) (internalId *internalResourceId, ref string, valid bool) {
+	idParts := strings.SplitN(altId, altIdSeparator, 2)
```

Comment:
It is up to us to define the format of Internal ID, so we should not include '|' character into internal ID, to avoid any confusions when parsing alternative ID.